core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/StepTracker.kt

```diff
@@ -0,0 +1,120 @@
+package fr.sncf.osrd.stdcm.infra_exploration
+
+import fr.sncf.osrd.graph.PathfindingEdgeLocationId
+import fr.sncf.osrd.sim_infra.api.Block
+import fr.sncf.osrd.sim_infra.api.BlockId
+import fr.sncf.osrd.sim_infra.api.TravelledPath
+import fr.sncf.osrd.stdcm.STDCMStep
+import fr.sncf.osrd.utils.units.Offset
+
+/**
+ * Component of `InfraExplorer` in charge of keeping track of anything related to the input steps.
+ * It keeps tracks of which steps we've passed and where, how many steps have been reached
+ * (specifically when we've reached the destination), and which steps are present in a block range.
+ *
+ * The tricky part is that we need to keep track of which steps are seen on the path as a whole
+ * (including lookahead), and which have actually been reached by the simulations (excluding
+ * lookahead).
+ *
+ * Unless specified otherwise, fields and methods refer to the whole path (including lookahead).
+ */
+class StepTracker(private val inputSteps: List<STDCMStep>) {
+    // If copying this is too expensive, it may be changed to an AppendOnlyLinkedList.
+    // But the list should be small.
+    private val seenSteps: MutableList<LocatedStep> = mutableListOf()
+    private val nSeenSteps: Int
+        get() = seenSteps.size
+
+    var nStepsExcludingLookahead: Int = 0
+        private set
+
+    // Used to compute path offsets
+    private var currentPathOffset: Offset<TravelledPath> = Offset.zero()
+
+    /** Returns all the steps that have been passed on the path, in order. */
+    fun getAllReachedSteps(): List<LocatedStep> {
```

Comment:
Not sure about the "All":
```suggestion
    fun getAllSeenSteps(): List<LocatedStep> {
```